src/backend/winit/mod.rs

```diff
@@ -372,234 +281,390 @@ where
+    NotSupported,
+    /// EGL error
+    #[error("EGL error: {0}")]
+    Egl(#[from] EGLError),
+    /// Renderer initialization failed
+    #[error("Renderer creation failed: {0}")]
+    RendererCreationError(#[from] GlesError),
+}
+
+/// Specific events generated by Winit
+#[derive(Debug)]
+pub enum WinitEvent {
+    /// The window has been resized
+    Resized {
+        /// The new physical size (in pixels)
+        size: Size<i32, Physical>,
+        /// The new scale factor
+        scale_factor: f64,
+    },
+
+    /// The focus state of the window changed
+    Focus(bool),
+
+    /// An input event occurred.
+    Input(InputEvent<WinitInput>),
+
+    /// A redraw was requested
+    Refresh,
+
+    /// The system manager asks to close the window.
+    CloseRequsted,
+
+    /// The winit loop was terimanted.
+    Terminated,
+}
+
+// TODO refactor to plug into calloop directly, once winit has the API to do so.
+fn winit_worker_start(
+    tx: CalloopSender<WinitEvent>,
+    window_tx: Sender<Result<(Arc<WinitWindow>, EventLoopProxy<WinitUserEvent>), Error>>,
+) {
+    let mut event_loop = match EventLoopBuilder::<WinitUserEvent>::with_user_event()
+        .with_any_thread(true)
+        .build()
+    {
+        Ok(event_loop) => event_loop,
+        Err(err) => {
+            let _ = window_tx.send(Err(Error::EventLoopError(err.to_string())));
+            return;
+        }
+    };
+
+    let window = match WindowBuilder::new().with_title("Smithay").build(&event_loop) {
+        Ok(window) => window,
+        Err(err) => {
+            let _ = window_tx.send(Err(Error::WindowCreationError(err.to_string())));
+            return;
+        }
+    };
+
+    let window = Arc::new(window);
+    let _ = window_tx.send(Ok((window.clone(), event_loop.create_proxy())));
+
+    let timestamp = {
+        let start_time = Instant::now();
+        move || start_time.elapsed().as_micros() as u64
+    };
+
+    let mut key_counter = 0u32;
+    let mut current_size = window.inner_size();
+    let mut current_scale_factor = window.scale_factor();
+    let is_x11 = matches!(event_loop.raw_display_handle(), RawDisplayHandle::Xlib(_));
+
+    loop {
+        event_loop.set_control_flow(ControlFlow::Wait);
+        let status = event_loop.pump_events(None, |event, window_target| {
+            match event {
+                Event::Resumed => {
+                    let _ = tx.send(WinitEvent::Input(InputEvent::DeviceAdded {
+                        device: WinitVirtualDevice,
+                    }));
+                }
+                Event::UserEvent(event) => match event {
+                    WinitUserEvent::Shutdown => window_target.exit(),
+                },
+                Event::WindowEvent { event, .. } => match event {
+                    WindowEvent::Resized(size) => {
+                        trace!("Resizing window to {size:?}");
+                        current_size = size;
```

Comment:
If the window is resized without a Resized event (i.e. with request_inner_size()) then this is not updated, messing up the input.

Reply:
I guess we could just wrap the way you set the `inner_size`, otherwise we'd have to go back to calling `inner_size` quite frequently, which is not that fast.

Reply:
Alternatively, we could simply wrap the way you get the `&WinitWindow` and provide it via some callback and then will just ensure the size/state of the window after each call to the `&WinitWindow`.